File: converted/source2/source/zombie/iso/IsoMarkers.java
package zombie.iso;

import java.util.ArrayList;
import java.util.List;
import se.krka.kahlua.vm.KahluaTable;
import zombie.GameTime;
import zombie.characters.IsoPlayer;
import zombie.core.SpriteRenderer;
import zombie.core.textures.Texture;
import zombie.debug.LineDrawer;
import zombie.iso.IsoCell.PerPlayerRender;
import zombie.iso.IsoMarkers.CircleIsoMarker;
import zombie.iso.IsoMarkers.IsoMarker;
import zombie.network.GameServer;

public final class IsoMarkers {
   public static final IsoMarkers instance = new IsoMarkers();
   private static int NextIsoMarkerID = 0;
   private final List<IsoMarker> markers = new ArrayList<>();
   private final List<CircleIsoMarker> circlemarkers = new ArrayList<>();
   private static int NextCircleIsoMarkerID = 0;

   private IsoMarkers() {
   }

   public void init() {
   }

   public void reset() {
      this.markers.clear();
      this.circlemarkers.clear();
   }

   public void update() {
      if (!GameServer.bServer) {
         this.updateIsoMarkers();
         this.updateCircleIsoMarkers();
      }
   }

   private void updateIsoMarkers() {
      if (IsoCamera.frameState.playerIndex == 0) {
         if (this.markers.size() != 0) {
            for (int var1 = this.markers.size() - 1; var1 >= 0; var1--) {
               if (this.markers.get(var1).isRemoved()) {
                  if (this.markers.get(var1).hasTempSquareObject()) {
                     this.markers.get(var1).removeTempSquareObjects();
                  }

                  this.markers.remove(var1);
               }
            }

            for (int var3 = 0; var3 < this.markers.size(); var3++) {
               IsoMarker var2 = this.markers.get(var3);
               if (var2.alphaInc) {
                  var2.alpha = var2.alpha + GameTime.getInstance().getMultiplier() * var2.fadeSpeed;
                  if (var2.alpha > var2.alphaMax) {
                     var2.alphaInc = false;
                     var2.alpha = var2.alphaMax;
                  }
               } else {
                  var2.alpha = var2.alpha - GameTime.getInstance().getMultiplier() * var2.fadeSpeed;
                  if (var2.alpha < var2.alphaMin) {
                     var2.alphaInc = true;
                     var2.alpha = 0.3F;
                  }
               }
            }
         }
      }
   }

   public boolean removeIsoMarker(IsoMarker var1) {
      return this.removeIsoMarker(var1.getID());
   }

   public boolean removeIsoMarker(int var1) {
      for (int var2 = this.markers.size() - 1; var2 >= 0; var2--) {
         if (this.markers.get(var2).getID() == var1) {
            this.markers.get(var2).remove();
            this.markers.remove(var2);
            return true;
         }
      }

      return false;
   }

   public IsoMarker getIsoMarker(int var1) {
      for (int var2 = 0; var2 < this.markers.size(); var2++) {
         if (this.markers.get(var2).getID() == var1) {
            return this.markers.get(var2);
         }
      }

      return null;
   }

   public IsoMarker addIsoMarker(String var1, IsoGridSquare var2, float var3, float var4, float var5, boolean var6, boolean var7) {
      if (GameServer.bServer) {
         return null;
      } else {
         IsoMarker var8 = new IsoMarker();
         var8.setSquare(var2);
         var8.init(var1, var2.x, var2.y, var2.z, var2, var7);
         var8.setR(var3);
         var8.setG(var4);
         var8.setB(var5);
         var8.setA(1.0F);
         var8.setDoAlpha(var6);
         var8.setFadeSpeed(0.006F);
         var8.setAlpha(1.0F);
         var8.setAlphaMin(0.3F);
         var8.setAlphaMax(1.0F);
         this.markers.add(var8);
         return var8;
      }
   }

   public IsoMarker addIsoMarker(KahluaTable var1, KahluaTable var2, IsoGridSquare var3, float var4, float var5, float var6, boolean var7, boolean var8) {
      return this.addIsoMarker(var1, var2, var3, var4, var5, var6, var7, var8, 0.006F, 0.3F, 1.0F);
   }

   public IsoMarker addIsoMarker(
      KahluaTable var1,
      KahluaTable var2,
      IsoGridSquare var3,
      float var4,
      float var5,
      float var6,
      boolean var7,
      boolean var8,
      float var9,
      float var10,
      float var11
   ) {
      if (GameServer.bServer) {
         return null;
      } else {
         IsoMarker var12 = new IsoMarker();
         var12.init(var1, var2, var3.x, var3.y, var3.z, var3, var8);
         var12.setSquare(var3);
         var12.setR(var4);
         var12.setG(var5);
         var12.setB(var6);
         var12.setA(1.0F);
         var12.setDoAlpha(var7);
         var12.setFadeSpeed(var9);
         var12.setAlpha(0.0F);
         var12.setAlphaMin(var10);
         var12.setAlphaMax(var11);
         this.markers.add(var12);
         return var12;
      }
   }

   public void renderIsoMarkers(PerPlayerRender var1, int var2, int var3) {
      if (!GameServer.bServer && this.markers.size() != 0) {
         IsoPlayer var4 = IsoPlayer.players[var3];
         if (var4 != null) {
            for (int var5 = 0; var5 < this.markers.size(); var5++) {
               IsoMarker var6 = this.markers.get(var5);
               if (var6.z == var2 && var6.z == var4.getZ() && var6.active) {
                  for (int var7 = 0; var7 < var6.textures.size(); var7++) {
                     Texture var8 = (Texture)var6.textures.get(var7);
                     float var9 = IsoUtils.XToScreen(var6.x, var6.y, var6.z, 0) - IsoCamera.cameras[var3].getOffX() - var8.getWidth() / 2.0F;
                     float var10 = IsoUtils.YToScreen(var6.x, var6.y, var6.z, 0) - IsoCamera.cameras[var3].getOffY() - var8.getHeight();
                     SpriteRenderer.instance.render(var8, var9, var10, var8.getWidth(), var8.getHeight(), var6.r, var6.g, var6.b, var6.alpha, null);
                  }
               }
            }
         }
      }
   }

   public void renderIsoMarkersDeferred(PerPlayerRender var1, int var2, int var3) {
      if (!GameServer.bServer && this.markers.size() != 0) {
         IsoPlayer var4 = IsoPlayer.players[var3];
         if (var4 != null) {
            for (int var5 = 0; var5 < this.markers.size(); var5++) {
               IsoMarker var6 = this.markers.get(var5);
               if (var6.z == var2 && var6.z == var4.getZ() && var6.active) {
                  for (int var7 = 0; var7 < var6.overlayTextures.size(); var7++) {
                     Texture var8 = (Texture)var6.overlayTextures.get(var7);
                     float var9 = IsoUtils.XToScreen(var6.x, var6.y, var6.z, 0) - IsoCamera.cameras[var3].getOffX() - var8.getWidth() / 2.0F;
                     float var10 = IsoUtils.YToScreen(var6.x, var6.y, var6.z, 0) - IsoCamera.cameras[var3].getOffY() - var8.getHeight();
                     SpriteRenderer.instance.render(var8, var9, var10, var8.getWidth(), var8.getHeight(), var6.r, var6.g, var6.b, var6.alpha, null);
                  }
               }
            }
         }
      }
   }

   public void renderIsoMarkersOnSquare(PerPlayerRender var1, int var2, int var3) {
      if (!GameServer.bServer && this.markers.size() != 0) {
         IsoPlayer var4 = IsoPlayer.players[var3];
         if (var4 != null) {
            for (int var5 = 0; var5 < this.markers.size(); var5++) {
               IsoMarker var6 = this.markers.get(var5);
               if (var6.z == var2 && var6.z == var4.getZ() && var6.active) {
                  for (int var7 = 0; var7 < var6.overlayTextures.size(); var7++) {
                     Texture var8 = (Texture)var6.overlayTextures.get(var7);
                     float var9 = IsoUtils.XToScreen(var6.x, var6.y, var6.z, 0) - IsoCamera.cameras[var3].getOffX() - var8.getWidth() / 2.0F;
                     float var10 = IsoUtils.YToScreen(var6.x, var6.y, var6.z, 0) - IsoCamera.cameras[var3].getOffY() - var8.getHeight();
                     SpriteRenderer.instance.render(var8, var9, var10, var8.getWidth(), var8.getHeight(), var6.r, var6.g, var6.b, var6.alpha, null);
                  }
               }
            }
         }
      }
   }

   private void updateCircleIsoMarkers() {
      if (IsoCamera.frameState.playerIndex == 0) {
         if (this.circlemarkers.size() != 0) {
            for (int var1 = this.circlemarkers.size() - 1; var1 >= 0; var1--) {
               if (this.circlemarkers.get(var1).isRemoved()) {
                  this.circlemarkers.remove(var1);
               }
            }

            for (int var3 = 0; var3 < this.circlemarkers.size(); var3++) {
               CircleIsoMarker var2 = this.circlemarkers.get(var3);
               if (var2.alphaInc) {
                  var2.alpha = var2.alpha + GameTime.getInstance().getMultiplier() * var2.fadeSpeed;
                  if (var2.alpha > var2.alphaMax) {
                     var2.alphaInc = false;
                     var2.alpha = var2.alphaMax;
                  }
               } else {
                  var2.alpha = var2.alpha - GameTime.getInstance().getMultiplier() * var2.fadeSpeed;
                  if (var2.alpha < var2.alphaMin) {
                     var2.alphaInc = true;
                     var2.alpha = 0.3F;
                  }
               }
            }
         }
      }
   }

   public boolean removeCircleIsoMarker(CircleIsoMarker var1) {
      return this.removeCircleIsoMarker(var1.getID());
   }

   public boolean removeCircleIsoMarker(int var1) {
      for (int var2 = this.circlemarkers.size() - 1; var2 >= 0; var2--) {
         if (this.circlemarkers.get(var2).getID() == var1) {
            this.circlemarkers.get(var2).remove();
            this.circlemarkers.remove(var2);
            return true;
         }
      }

      return false;
   }

   public CircleIsoMarker getCircleIsoMarker(int var1) {
      for (int var2 = 0; var2 < this.circlemarkers.size(); var2++) {
         if (this.circlemarkers.get(var2).getID() == var1) {
            return this.circlemarkers.get(var2);
         }
      }

      return null;
   }

   public CircleIsoMarker addCircleIsoMarker(IsoGridSquare var1, float var2, float var3, float var4, float var5) {
      if (GameServer.bServer) {
         return null;
      } else {
         CircleIsoMarker var6 = new CircleIsoMarker();
         var6.init(var1.x, var1.y, var1.z, var1);
         var6.setSquare(var1);
         var6.setR(var2);
         var6.setG(var3);
         var6.setB(var4);
         var6.setA(var5);
         var6.setDoAlpha(false);
         var6.setFadeSpeed(0.006F);
         var6.setAlpha(1.0F);
         var6.setAlphaMin(1.0F);
         var6.setAlphaMax(1.0F);
         this.circlemarkers.add(var6);
         return var6;
      }
   }

   public void renderCircleIsoMarkers(PerPlayerRender var1, int var2, int var3) {
      if (!GameServer.bServer && this.circlemarkers.size() != 0) {
         IsoPlayer var4 = IsoPlayer.players[var3];
         if (var4 != null) {
            for (int var5 = 0; var5 < this.circlemarkers.size(); var5++) {
               CircleIsoMarker var6 = this.circlemarkers.get(var5);
               if (var6.z == var2 && var6.z == var4.getZ() && var6.active) {
                  LineDrawer.DrawIsoCircle(var6.x, var6.y, var6.z, var6.size, 32, var6.r, var6.g, var6.b, var6.a);
               }
            }
         }
      }
   }

   public void render() {
      this.update();
   }
}

File: ProjectZomboidSwitch/src/zombie/iso/IsoMarkers.cpp
// --- IsoMarker methods ---
// (Assumes IsoMarker is now a nested class in IsoMarkers.h)
namespace zombie { namespace iso {

int IsoMarkers::IsoMarker::getID() const { return ID; }
void IsoMarkers::IsoMarker::remove() { removed = true; }
bool IsoMarkers::IsoMarker::isRemoved() const { return removed; }
void IsoMarkers::IsoMarker::init(LuaTable* t1, LuaTable* t2, int px, int py, int pz, std::shared_ptr<IsoGridSquare> sq) { square = sq; setPos(px, py, pz); /* TODO: Fill textures/overlayTextures from LuaTable */ }
void IsoMarkers::IsoMarker::init(LuaTable* t1, LuaTable* t2, int px, int py, int pz, std::shared_ptr<IsoGridSquare> sq, bool temp) { square = sq; setPos(px, py, pz); /* TODO: Fill tempObjects from LuaTable if temp */ }
void IsoMarkers::IsoMarker::init(const std::string& name, int px, int py, int pz, std::shared_ptr<IsoGridSquare> sq, bool temp) { square = sq; setPos(px, py, pz); /* TODO: Fill tempObjects if temp */ }
bool IsoMarkers::IsoMarker::hasTempSquareObject() const { return !tempObjects.empty(); }
void IsoMarkers::IsoMarker::addTempSquareObject(std::shared_ptr<IsoObject> obj) { /* TODO: Add to square->localTemporaryObjects */ }
void IsoMarkers::IsoMarker::removeTempSquareObjects() { /* TODO: Remove from square->localTemporaryObjects */ }
float IsoMarkers::IsoMarker::getX() const { return x; }
float IsoMarkers::IsoMarker::getY() const { return y; }
float IsoMarkers::IsoMarker::getZ() const { return z; }
float IsoMarkers::IsoMarker::getR() const { return r; }
float IsoMarkers::IsoMarker::getG() const { return g; }
float IsoMarkers::IsoMarker::getB() const { return b; }
float IsoMarkers::IsoMarker::getA() const { return a; }
void IsoMarkers::IsoMarker::setR(float v) { r = v; }
void IsoMarkers::IsoMarker::setG(float v) { g = v; }
void IsoMarkers::IsoMarker::setB(float v) { b = v; }
void IsoMarkers::IsoMarker::setA(float v) { a = v; }
float IsoMarkers::IsoMarker::getAlpha() const { return alpha; }
void IsoMarkers::IsoMarker::setAlpha(float v) { alpha = v; }
float IsoMarkers::IsoMarker::getAlphaMax() const { return alphaMax; }
void IsoMarkers::IsoMarker::setAlphaMax(float v) { alphaMax = v; }
float IsoMarkers::IsoMarker::getAlphaMin() const { return alphaMin; }
void IsoMarkers::IsoMarker::setAlphaMin(float v) { alphaMin = v; }
bool IsoMarkers::IsoMarker::isDoAlpha() const { return doAlpha; }
void IsoMarkers::IsoMarker::setDoAlpha(bool v) { doAlpha = v; }
float IsoMarkers::IsoMarker::getFadeSpeed() const { return fadeSpeed; }
void IsoMarkers::IsoMarker::setFadeSpeed(float v) { fadeSpeed = v; }
std::shared_ptr<IsoGridSquare> IsoMarkers::IsoMarker::getSquare() const { return square; }
void IsoMarkers::IsoMarker::setSquare(std::shared_ptr<IsoGridSquare> sq) { square = sq; }
void IsoMarkers::IsoMarker::setPos(int px, int py, int pz) { x = px + 0.5f; y = py + 0.5f; z = pz; }
bool IsoMarkers::IsoMarker::isActive() const { return active; }
void IsoMarkers::IsoMarker::setActive(bool v) { active = v; }

// --- CircleIsoMarker methods ---
int IsoMarkers::CircleIsoMarker::getID() const { return ID; }
void IsoMarkers::CircleIsoMarker::remove() { removed = true; }
bool IsoMarkers::CircleIsoMarker::isRemoved() const { return removed; }
void IsoMarkers::CircleIsoMarker::init(int px, int py, int pz, std::shared_ptr<IsoGridSquare> sq) { square = sq; setPos(px, py, pz); }
float IsoMarkers::CircleIsoMarker::getX() const { return x; }
float IsoMarkers::CircleIsoMarker::getY() const { return y; }
float IsoMarkers::CircleIsoMarker::getZ() const { return z; }
float IsoMarkers::CircleIsoMarker::getR() const { return r; }
float IsoMarkers::CircleIsoMarker::getG() const { return g; }
float IsoMarkers::CircleIsoMarker::getB() const { return b; }
float IsoMarkers::CircleIsoMarker::getA() const { return a; }
void IsoMarkers::CircleIsoMarker::setR(float v) { r = v; }
void IsoMarkers::CircleIsoMarker::setG(float v) { g = v; }
void IsoMarkers::CircleIsoMarker::setB(float v) { b = v; }
void IsoMarkers::CircleIsoMarker::setA(float v) { a = v; }
float IsoMarkers::CircleIsoMarker::getSize() const { return size; }
void IsoMarkers::CircleIsoMarker::setSize(float v) { size = v; }
float IsoMarkers::CircleIsoMarker::getAlpha() const { return alpha; }
void IsoMarkers::CircleIsoMarker::setAlpha(float v) { alpha = v; }
float IsoMarkers::CircleIsoMarker::getAlphaMax() const { return alphaMax; }
void IsoMarkers::CircleIsoMarker::setAlphaMax(float v) { alphaMax = v; }
float IsoMarkers::CircleIsoMarker::getAlphaMin() const { return alphaMin; }
void IsoMarkers::CircleIsoMarker::setAlphaMin(float v) { alphaMin = v; }
bool IsoMarkers::CircleIsoMarker::isDoAlpha() const { return doAlpha; }
void IsoMarkers::CircleIsoMarker::setDoAlpha(bool v) { doAlpha = v; }
float IsoMarkers::CircleIsoMarker::getFadeSpeed() const { return fadeSpeed; }
void IsoMarkers::CircleIsoMarker::setFadeSpeed(float v) { fadeSpeed = v; }
std::shared_ptr<IsoGridSquare> IsoMarkers::CircleIsoMarker::getSquare() const { return square; }
void IsoMarkers::CircleIsoMarker::setSquare(std::shared_ptr<IsoGridSquare> sq) { square = sq; }
void IsoMarkers::CircleIsoMarker::setPos(int px, int py, int pz) { x = px; y = py; z = pz; }
bool IsoMarkers::CircleIsoMarker::isActive() const { return active; }
void IsoMarkers::CircleIsoMarker::setActive(bool v) { active = v; }

} // namespace iso } // namespace zombie

#include <string>
#include "zombie/iso/IsoMarkers.h"

namespace zombie {
namespace iso {

IsoMarkers::IsoMarkers() {}

void IsoMarkers::init() {}

void IsoMarkers::reset() {
    markers.clear();
    circlemarkers.clear();
}

void IsoMarkers::update() {
    if (!GameServer::bServer) {
        updateIsoMarkers();
        updateCircleIsoMarkers();
    }
}

void IsoMarkers::updateIsoMarkers() {
    if (IsoCamera::frameState.playerIndex == 0) {
        if (!markers.empty()) {
            for (int i = static_cast<int>(markers.size()) - 1; i >= 0; --i) {
                if (markers[i]->isRemoved()) {
                    if (markers[i]->hasTempSquareObject()) {
                        markers[i]->removeTempSquareObjects();
                    }
                    markers.erase(markers.begin() + i);
                }
            }
            for (auto& marker : markers) {
                if (marker->alphaInc) {
                    marker->alpha += GameTime::getInstance()->getMultiplier() * marker->fadeSpeed;
                    if (marker->alpha > marker->alphaMax) {
                        marker->alphaInc = false;
                        marker->alpha = marker->alphaMax;
                    }
                } else {
                    marker->alpha -= GameTime::getInstance()->getMultiplier() * marker->fadeSpeed;
                    if (marker->alpha < marker->alphaMin) {
                        marker->alphaInc = true;
                        marker->alpha = 0.3f;
                    }
                }
            }
        }
    }
}

bool IsoMarkers::removeIsoMarker(std::shared_ptr<IsoMarker> marker) {
    return removeIsoMarker(marker->getID());
}

bool IsoMarkers::removeIsoMarker(int id) {
    for (int i = static_cast<int>(markers.size()) - 1; i >= 0; --i) {
        if (markers[i]->getID() == id) {
            markers[i]->remove();
            markers.erase(markers.begin() + i);
            return true;
        }
    }
    return false;
}

std::shared_ptr<IsoMarker> IsoMarkers::getIsoMarker(int id) {
    for (auto& marker : markers) {
        if (marker->getID() == id) return marker;
    }
    return nullptr;
}

std::shared_ptr<IsoMarker> IsoMarkers::addIsoMarker(const std::string& name, std::shared_ptr<IsoGridSquare> sq, float r, float g, float b, bool doAlpha, bool temp) {
    if (GameServer::bServer) return nullptr;
    auto marker = std::make_shared<IsoMarker>();
    marker->setSquare(sq);
    marker->init(name, sq->x, sq->y, sq->z, sq, temp);
    marker->setR(r);
    marker->setG(g);
    marker->setB(b);
    marker->setA(1.0f);
    marker->setDoAlpha(doAlpha);
    marker->setFadeSpeed(0.006f);
    marker->setAlpha(1.0f);
    marker->setAlphaMin(0.3f);
    marker->setAlphaMax(1.0f);
    markers.push_back(marker);
    return marker;
}

std::shared_ptr<IsoMarker> IsoMarkers::addIsoMarker(/*LuaTable*/void* t1, /*LuaTable*/void* t2, std::shared_ptr<IsoGridSquare> sq, float r, float g, float b, bool doAlpha, bool temp) {
    // LuaTable types are placeholders; replace with your LuaTable wrapper if needed
    return addIsoMarker(t1, t2, sq, r, g, b, doAlpha, temp, 0.006f, 0.3f, 1.0f);
}

std::shared_ptr<IsoMarker> IsoMarkers::addIsoMarker(/*LuaTable*/void* t1, /*LuaTable*/void* t2, std::shared_ptr<IsoGridSquare> sq, float r, float g, float b, bool doAlpha, bool temp, float fadeSpeed, float alphaMin, float alphaMax) {
    if (GameServer::bServer) return nullptr;
    auto marker = std::make_shared<IsoMarker>();
    marker->init(t1, t2, sq->x, sq->y, sq->z, sq, temp);
    marker->setSquare(sq);
    marker->setR(r);
    marker->setG(g);
    marker->setB(b);
    marker->setA(1.0f);
    marker->setDoAlpha(doAlpha);
    marker->setFadeSpeed(fadeSpeed);
    marker->setAlpha(0.0f);
    marker->setAlphaMin(alphaMin);
    marker->setAlphaMax(alphaMax);
    markers.push_back(marker);
    return marker;
}

void IsoMarkers::renderIsoMarkers(std::shared_ptr<PerPlayerRender> render, int z, int playerIdx) {
    if (!GameServer::bServer && !markers.empty()) {
        auto player = IsoPlayer::players[playerIdx];
        if (player) {
            for (auto& marker : markers) {
                if (marker->z == z && marker->z == player->getZ() && marker->active) {
                    for (auto& tex : marker->textures) {
                        float sx = IsoUtils::XToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffX() - tex->getWidth() / 2.0f;
                        float sy = IsoUtils::YToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffY() - tex->getHeight();
                        SpriteRenderer::instance->render(tex, sx, sy, tex->getWidth(), tex->getHeight(), marker->r, marker->g, marker->b, marker->alpha, nullptr);
                    }
                }
            }
        }
    }
}

void IsoMarkers::renderIsoMarkersDeferred(std::shared_ptr<PerPlayerRender> render, int z, int playerIdx) {
    if (!GameServer::bServer && !markers.empty()) {
        auto player = IsoPlayer::players[playerIdx];
        if (player) {
            for (auto& marker : markers) {
                if (marker->z == z && marker->z == player->getZ() && marker->active) {
                    for (auto& tex : marker->overlayTextures) {
                        float sx = IsoUtils::XToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffX() - tex->getWidth() / 2.0f;
                        float sy = IsoUtils::YToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffY() - tex->getHeight();
                        SpriteRenderer::instance->render(tex, sx, sy, tex->getWidth(), tex->getHeight(), marker->r, marker->g, marker->b, marker->alpha, nullptr);
                    }
                }
            }
        }
    }
}

// Add updateCircleIsoMarkers and renderIsoMarkersOnSquare as needed for completeness

} // namespace iso
} // namespace zombie


namespace zombie {
namespace iso {

void IsoMarkers::renderIsoMarkersOnSquare(std::shared_ptr<PerPlayerRender> render, int z, int playerIdx) {
    if (!GameServer::bServer && !markers.empty()) {
        auto player = IsoPlayer::players[playerIdx];
        if (player) {
            for (auto& marker : markers) {
                if (marker->z == z && marker->z == player->getZ() && marker->active) {
                    for (auto& tex : marker->overlayTextures) {
                        float sx = IsoUtils::XToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffX() - tex->getWidth() / 2.0f;
                        float sy = IsoUtils::YToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffY() - tex->getHeight();
                        SpriteRenderer::instance->render(tex, sx, sy, tex->getWidth(), tex->getHeight(), marker->r, marker->g, marker->b, marker->alpha, nullptr);
                    }
                }
            }
        }
    }
}

void IsoMarkers::updateCircleIsoMarkers() {
    if (IsoCamera::frameState.playerIndex == 0) {
        if (!circlemarkers.empty()) {
            for (int i = static_cast<int>(circlemarkers.size()) - 1; i >= 0; --i) {
                if (circlemarkers[i]->isRemoved()) {
                    circlemarkers.erase(circlemarkers.begin() + i);
                }
            }
            for (auto& marker : circlemarkers) {
                if (marker->alphaInc) {
                    marker->alpha += GameTime::getInstance()->getMultiplier() * marker->fadeSpeed;
                    if (marker->alpha > marker->alphaMax) {
                        marker->alphaInc = false;
                        marker->alpha = marker->alphaMax;
                    }
                } else {
                    marker->alpha -= GameTime::getInstance()->getMultiplier() * marker->fadeSpeed;
                    if (marker->alpha < marker->alphaMin) {
                        marker->alphaInc = true;
                        marker->alpha = 0.3f;
                    }
                }
            }
        }
    }
}

bool IsoMarkers::removeCircleIsoMarker(std::shared_ptr<CircleIsoMarker> marker) {
    for (int i = static_cast<int>(circlemarkers.size()) - 1; i >= 0; --i) {
        if (circlemarkers[i] == marker) {
            circlemarkers.erase(circlemarkers.begin() + i);
            return true;
        }
    }
    return false;
}

bool IsoMarkers::removeCircleIsoMarker(int id) {
    for (int i = static_cast<int>(circlemarkers.size()) - 1; i >= 0; --i) {
        if (circlemarkers[i]->getID() == id) {
            circlemarkers.erase(circlemarkers.begin() + i);
            return true;
        }
    }
    return false;
}

std::shared_ptr<CircleIsoMarker> IsoMarkers::getCircleIsoMarker(int id) {
    for (auto& marker : circlemarkers) {
        if (marker->getID() == id) return marker;
    }
    return nullptr;
}

std::shared_ptr<CircleIsoMarker> IsoMarkers::addCircleIsoMarker(std::shared_ptr<IsoGridSquare> sq, float r, float g, float b, float a) {
    if (GameServer::bServer) return nullptr;
    auto marker = std::make_shared<CircleIsoMarker>();
    marker->setSquare(sq);
    marker->setR(r);
    marker->setG(g);
    marker->setB(b);
    marker->setA(a);
    marker->setFadeSpeed(0.006f);
    marker->setAlpha(1.0f);
    marker->setAlphaMin(0.3f);
    marker->setAlphaMax(1.0f);
    circlemarkers.push_back(marker);
    return marker;
}

void IsoMarkers::renderCircleIsoMarkers(std::shared_ptr<PerPlayerRender> render, int z, int playerIdx) {
    if (!GameServer::bServer && !circlemarkers.empty()) {
        auto player = IsoPlayer::players[playerIdx];
        if (player) {
            for (auto& marker : circlemarkers) {
                if (marker->z == z && marker->z == player->getZ() && marker->active) {
                    for (auto& tex : marker->textures) {
                        float sx = IsoUtils::XToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffX() - tex->getWidth() / 2.0f;
                        float sy = IsoUtils::YToScreen(marker->x, marker->y, marker->z, 0) - IsoCamera::cameras[playerIdx]->getOffY() - tex->getHeight();
                        SpriteRenderer::instance->render(tex, sx, sy, tex->getWidth(), tex->getHeight(), marker->r, marker->g, marker->b, marker->alpha, nullptr);
                    }
                }
            }
        }
    }
}

void IsoMarkers::render() {
    if (GameServer::bServer) return;
    int numPlayers = static_cast<int>(IsoPlayer::players.size());
    for (int playerIdx = 0; playerIdx < numPlayers; ++playerIdx) {
        auto player = IsoPlayer::players[playerIdx];
        if (!player) continue;
        int z = player->getZ();
        auto render = IsoCamera::cameras[playerIdx];
        // Render standard markers
        renderIsoMarkers(render, z, playerIdx);
        // Render deferred/overlay markers
        renderIsoMarkersDeferred(render, z, playerIdx);
        // Render markers on square (if needed)
        renderIsoMarkersOnSquare(render, z, playerIdx);
        // Render circle markers
        renderCircleIsoMarkers(render, z, playerIdx);
    }
}

} // namespace iso
} // namespace zombie

